src/internal/pachd/base.go

```diff
@@ -0,0 +1,90 @@
+package pachd
+
+import (
+	"context"
+	"fmt"
+	"runtime/debug"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachconfig"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/version"
+	"go.uber.org/automaxprocs/maxprocs"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type setupStep struct {
+	Name string
+	Fn   func(context.Context) error
+}
+
+// base structures the setup and run phases of pachds
+type base struct {
+	setup      []setupStep
+	background map[string]func(context.Context) error
+	done       bool
+
+	config pachconfig.Configuration
```

Comment:
I think it would be nice if this was a `type base[T isPachdConfig] { ... config: T }` and the associated tweaks throughout.  cmdutil.Main is already this type, though it uses any instead of `isPachdConfig`.

That would be `type isPachdConfig interface { isPachdConfig() }` and then in config `func (*PreflightConfiguration) isPachdConfig() {} `.

That way, nobody can misuse cmdutil or base, and it doesn't cause the anti-pattern of including more config than you need.

Reply:
pachctl-doc creates it's own `appEnv` which is the immediate problem with this.

The purpose of the config was just to call tweakResources, which is probably fine to leave out for pre-flight.  I think in general, pachds will compose the base struct and then add the config themselves.

Reply:
OK.  The state right now is that a lot of things take config that they don't need, and they complain if it's unset.  preflight was the first not to do that.  The helm chart has to be updated to provide fake values for other required things, if I recall correctly.